source/common/protobuf/deterministic_hash.cc

```diff
@@ -0,0 +1,309 @@
+    break;
+  case FieldDescriptor::CPPTYPE_UINT32:
+    seed = hashScalarField<uint32_t>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_INT64:
+    seed = hashScalarField<int64_t>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_UINT64:
+    seed = hashScalarField<uint64_t>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_DOUBLE:
+    seed = hashScalarField<double>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_FLOAT:
+    seed = hashScalarField<float>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_BOOL:
+    seed = hashScalarField<bool>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_ENUM:
+    if (field.is_repeated()) {
+      int c = reflection.FieldSize(message, &field);
+      for (int i = 0; i < c; i++) {
+        int v = reflection.GetRepeatedEnumValue(message, &field, i);
+        seed = HashUtil::xxHash64(absl::string_view{reinterpret_cast<char*>(&v), sizeof(v)}, seed);
+      }
+    } else {
+      int v = reflection.GetEnumValue(message, &field);
+      seed = HashUtil::xxHash64(absl::string_view{reinterpret_cast<char*>(&v), sizeof(v)}, seed);
+    }
+    break;
+  case FieldDescriptor::CPPTYPE_STRING: {
+    if (field.is_repeated()) {
+      for (const std::string& str : reflection.GetRepeatedFieldRef<std::string>(message, &field)) {
+        seed = HashUtil::xxHash64(str, seed);
+      }
+    } else {
+      std::string scratch;
+      seed = HashUtil::xxHash64(reflection.GetStringReference(message, &field, &scratch), seed);
+    }
+  } break;
+  case FieldDescriptor::CPPTYPE_MESSAGE:
+    if (field.is_map()) {
+      seed = reflectionHashMapField(message, field, seed);
+    } else if (field.is_repeated()) {
+      for (const Protobuf::Message& submsg :
+           reflection.GetRepeatedFieldRef<Protobuf::Message>(message, &field)) {
+        seed = reflectionHashMessage(submsg, seed);
+      }
+    } else {
+      seed = reflectionHashMessage(reflection.GetMessage(message, &field), seed);
+    }
+    break;
+  }
+  return seed;
+}
+
+absl::string_view typeUrlToDescriptorFullName(absl::string_view url) {
+  const size_t pos = url.rfind('/');
+  if (pos != absl::string_view::npos) {
+    return url.substr(pos + 1);
+  }
+  return url;
+}
+
+std::unique_ptr<Protobuf::Message> unpackAnyForReflection(const ProtobufWkt::Any& any) {
+  const Protobuf::Descriptor* descriptor =
+      Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(
+          typeUrlToDescriptorFullName(any.type_url()));
+  if (descriptor == nullptr) {
+    return nullptr;
+  }
+  const Protobuf::Message* prototype =
+      Protobuf::MessageFactory::generated_factory()->GetPrototype(descriptor);
+  std::unique_ptr<Protobuf::Message> msg(prototype->New());
+  any.UnpackTo(msg.get());
+  return msg;
+}
+
+// This is intentionally ignoring unknown fields.
```

Comment:
Added to the header file comment (at a commit not pushed yet because I'm waiting for the CI coverage run before pushing more). MessageDifferencer can be configured to ignore or not ignore unknown fields. I don't think anything is actually making a hash table using the Utility::hash call right now but it's reasonable to comment about it in case of future usage.